Hw3

ViciousXerra opened the pull request.

ViciousXerra committed:
Started

ViciousXerra committed:
Bot RestController completed

ViciousXerra committed:
Scrapper requests/responses and error responses DTO created

ViciousXerra committed:
configuring springdoc

ViciousXerra committed:
trying to bind springdoc to port 8080

ViciousXerra committed:
added custom link validation
created documentation

ViciousXerra committed:
create client interfaces to scrapper api

ViciousXerra committed:
Update scrapper and bot clients
Created DTOs for clients

ViciousXerra committed:
Create config to scrapper clients

ViciousXerra committed:
Create config to bot client

ViciousXerra committed:
Refactoring

ViciousXerra committed:
???

ViciousXerra committed:
Why POST, GET /links legit and DELETE are not?

ViciousXerra committed:
Solved, fixed LinksClient.java

ViciousXerra committed:
Solved, fixed LinksClient.java

ViciousXerra committed:
Fixed checkstyle

ViciousXerra committed:
Fixed checkstyle

ViciousXerra committed:
Fix closing spring context

ViciousXerra committed:
Fix closing spring context

ViciousXerra committed:
Fix closing spring context

ViciousXerra committed:
Create BotClient tests

ViciousXerra committed:
Create BotClient tests

ViciousXerra committed:
Redacted scrapper client tests

ViciousXerra committed:
Redacted bot client tests

ViciousXerra committed:
Add BotRestController basic web mvc tests

ViciousXerra committed:
Add TelegramChatController basic web mvc tests

ViciousXerra committed:
Add TelegramChatController basic web mvc tests

ViciousXerra committed:
Tests redacted

ViciousXerra committed:
Tests redacted

Betorov commented on bot/src/main/java/edu/java/bot/api/controlleradvices/BotControllerAdvice.java:
```diff
@@ -0,0 +1,50 @@
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ResponseEntity<ApiErrorResponse> handleLinkUpdateBadRequest(MethodArgumentNotValidException e) {
+        return new ResponseEntity<>(
+            createApiErrorResponse("Invalid or incorrect request parameters", HttpStatus.BAD_REQUEST, e),
```
@ResponseStatus, и далее используешь этот же код в создании ответа ошибки читаемого, мне кажется оверхед. Из пушки по воробьям

Betorov commented on bot/src/main/java/edu/java/bot/api/dto/customvalidations/LinkValidation.java:
```diff
@@ -0,0 +1,23 @@
+@Target({ElementType.METHOD, ElementType.FIELD})
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+public @interface LinkValidation {
```
Вроде не очень большие классы, и можно было слить с LinkValidator (вложенность). Все равно раздельно не используют, а платят не за количество создаваемых классов

ViciousXerra replied:
Согласен, вложил аннотацию в сам валидатор.

Betorov commented on bot/src/main/java/edu/java/bot/api/dto/requests/LinkUpdate.java:
```diff
@@ -0,0 +1,17 @@
+public record LinkUpdate(
+    long id,
+    @LinkValidation
+    String url,
```
Тут бы не помещал @NotBlank

ViciousXerra replied:
Сделано

Betorov commented on bot/src/main/java/edu/java/bot/api/restcontrollers/BotRestController.java:
```diff
@@ -0,0 +1,50 @@
+    @Operation(summary = "Post a link update")
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "Link update was successfully processed"),
+        @ApiResponse(
+            responseCode = "400", description = "Invalid request body",
+            content = @Content(
+                mediaType = "application/json",
+                schema = @Schema(implementation = ApiErrorResponse.class)
+            )
+        ),
+        @ApiResponse(
+            responseCode = "409", description = "Link update already exists",
+            content = @Content(
+                mediaType = "application/json",
+                schema = @Schema(implementation = ApiErrorResponse.class)
+            )
+        )
```
@ResponseStatus уже реализовал в обработчике выше. 
Попробуй без этих строк запустить и сваггер будет такой же :)

Двойная работа, и джаве будет плохо, проблема выбора

ViciousXerra replied:
Да, swagger и без них работает, просто думал, что генерация дополнительной инфы по типу description из спецификации обязательна.

Betorov commented on bot/src/main/java/edu/java/bot/configuration/ScrapperClientConfig.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    private static boolean isInvalidBaseUrl(String url) {
+        return url == null || url.isEmpty() || url.isBlank();
```
Перечитал раза 3.

Нет меча или Нет рукояти меча или Лезвие тупое.

Мне кажется url.isBlank() достаточно

Betorov commented on bot/src/main/java/edu/java/bot/scrapperclient/clients/ChatClient.java:
```diff
@@ -0,0 +1,18 @@
+@HttpExchange(url = "/tg-chat")
+public interface ChatClient {
+
+    @PostExchange(url = "/{id}")
+    ResponseEntity<?> signUpChat(@PathVariable long id);
+
+    @DeleteExchange(url = "/{id}")
```
/{id} дублируются, надо бы вынести в словарик.

Учимся и вырабатываем полезные привычки

ViciousXerra replied:
Сделано

Betorov commented on bot/src/main/java/edu/java/bot/scrapperclient/clients/LinksClient.java:
```diff
@@ -0,0 +1,32 @@
+public interface LinksClient {
+
+    @GetExchange
+    ListLinkResponse getAllLinks(@RequestHeader("Tg-Chat-Id") long chatId);
+
+    @PostExchange
+    LinkResponse addLink(
+        @RequestHeader("Tg-Chat-Id") long chatId,
+        @RequestBody AddLinkRequest addLinkRequest
+    );
+
+    @DeleteExchange
+    LinkResponse removeLink(
+        @RequestHeader("Tg-Chat-Id") long chatId,
```
Tg-Chat-Id дублируется аж 3 раза!
:)

ViciousXerra replied:
Сделано

Betorov commented on bot/src/test/java/edu/java/bot/botrestcontrollerstests/BotRestControllerTest.java:
```diff
@@ -0,0 +1,94 @@
+            post("/bot/updates")
+                .contentType("application/json")
+                .content("""
+                    {
+                      "id": 1,
+                      "url": "https://github.com",
+                      "description": "desc",
+                      "tgChatIds": [
+                        1,
+                        2,
+                        3
+                      ]
+                    }""")
```
Давай лучше выше в константу вынесем. Сложно читается.

```
val actualBody = 
                            """
                            """
mockMvc...
```

Betorov commented on bot/src/test/java/edu/java/bot/botrestcontrollerstests/BotRestControllerTest.java:
```diff
@@ -0,0 +1,94 @@
+
+    private static String[] provideBadRequestBody() {
+        return new String[] {
+            """
+                    {
+                      "id": 1,
+                      "url": "not a link",
+                      "description": "desc",
+                      "tgChatIds": [
+                        1,
+                        2,
+                        3
+                      ]
+                    }""",
+            """
+                    {
+                      "id": 1,
+                      "url": "http://github.com",
+                      "description": "",
+                      "tgChatIds": [
+                        1,
+                        2,
+                        3
+                      ]
+                    }""",
+            """
+                    {
+                      "id": 1,
+                      "url": "http://github.com",
+                      "description": "desc",
+                      "tgChatIds": [
+                      ]
+                    }"""
```
Лучше делать так

```
"""
SOME_TEXT
"""
```
Копировать вставить и отправить будет гораздо легче, к примеру тестировщику чтоб понял как оно работает

ViciousXerra replied:
Подправил

Betorov commented on bot/src/test/java/edu/java/bot/scrapperclientstests/ChatClientTest.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Test
+    @Order(1)
```
Плохо прослеживается а это зачем?

ViciousXerra replied:
Я когда прошлый пр открывал, у гитхаба сломался workflow :). Пока сообразил проверить: "Точно ли на моей стороне проблема?", уже успел нарубить дров по типу того, что выше. Грешил на то, что контекст контейнера не закрывается.

ViciousXerra committed:
Refactoring:
Nested LinkValidation annotation in LinkValidator.java
Fixed @LinkValidation in DTOs
Fixed ScrapperClientConfig.java
Fixed BotControllerAdvice.java, ScrapperControllerAdvice.java
Removed "documentation" annotations in REST - controllers

ViciousXerra committed:
Refactoring:
ChatClient.java
LinksClient.java

ViciousXerra committed:
Removed dirties context annotation
Redacted ScrapperClientConfig.java

ViciousXerra committed:
Removed unused imports

ViciousXerra committed:
Tests redacted.

ViciousXerra committed:
Tests redacted.

ViciousXerra committed:
Tests redacted.

ViciousXerra committed:
Scrapper REST controllers refactoring

ViciousXerra committed:
Fixed BotClientConfig.java

ViciousXerra committed:
Merge branch 'main' into hw3

ViciousXerra committed:
merging main into hw3

ViciousXerra committed:
Merge remote-tracking branch 'java2024/hw3' into hw3

# Conflicts:
#	bot/src/main/resources/application.yml
#	scrapper/src/main/java/edu/java/scrapper/configuration/ApplicationConfig.java
